src/diffusers/pipelines/pipeline_loading_utils.py

```diff
@@ -358,6 +363,125 @@ def _get_pipeline_class(
+    is_diffusers_model = issubclass(class_obj, diffusers_module.ModelMixin)
+
+    model = None
+    config_path = cached_folder
+    user_agent = {
+        "diffusers": __version__,
+        "file_type": "model",
+        "framework": "pytorch",
+    }
+
+    if is_diffusers_model or is_transformers_model:
+        if hasattr(class_obj, "load_config"):
+            # Load config and then the model on meta.
+            config, unused_kwargs, commit_hash = class_obj.load_config(
+                os.path.join(config_path, name),
+                cache_dir=cached_folder,
+                return_unused_kwargs=True,
+                return_commit_hash=True,
+                force_download=kwargs.pop("force_download", False),
+                resume_download=kwargs.pop("resume_download", False),
+                proxies=kwargs.pop("proxies", None),
+                local_files_only=kwargs.pop("local_files_only", False),
+                token=kwargs.pop("token", None),
+                revision=kwargs.pop("revision", None),
+                subfolder=kwargs.pop("subfolder", None),
+                user_agent=user_agent,
+            )
+            with accelerate.init_empty_weights():
+                model = class_obj.from_config(config, **unused_kwargs)
+        else:
+            config_class = getattr(class_obj, "config_class", None)
+            if config_class is None:
+                raise ValueError("`config_class` cannot be None. Please double-check the model.")
+
+            config = config_class.from_pretrained(
+                cached_folder,
+                subfolder=name,
+                force_download=kwargs.pop("force_download", False),
+                resume_download=kwargs.pop("resume_download", False),
+                proxies=kwargs.pop("proxies", None),
+                local_files_only=kwargs.pop("local_files_only", False),
+                token=kwargs.pop("token", None),
+                revision=kwargs.pop("revision", None),
+                user_agent=user_agent,
+            )
+            with accelerate.init_empty_weights():
+                model = class_obj(config)
+
+    if model is not None:
+        model = model.to(dtype=torch_dtype)
+    return model
+
+
+def _assign_components_to_devices(
+    module_sizes: Dict[str, float], device_memory: Dict[str, float], device_mapping_strategy: str = "balanced"
+):
+    device_ids = list(device_memory.keys())
+    device_cycle = device_ids + device_ids[::-1]
+
+    deivce_id_component_mapping = {}
+    needs_offloading_mapping = {}
+    current_device_index = 0
+    for component in module_sizes:
+        device_id = device_cycle[current_device_index % len(device_cycle)]
+
+        component_memory = module_sizes[component]
+        curr_device_memory = device_memory[device_id]
+
+        # If the GPU doesn't fit the current component offload to the CPU.
+        if component_memory > curr_device_memory:
+            needs_offloading_mapping[component] = True
```

Comment:
We don't need this var since we know which model to offload by checking his device_map == "cpu"